### ISSUE TITLE
npm@5 ignores `"os": ["darwin"]` on Windows

### ISSUE THREAD

**Issue by refack** [npm/npm#16810]
#### I'm opening this issue because:

  - [X] npm is producing an incorrect install.

#### What's going wrong?
When I run `npm i fsevents@0.3.4` on windows, it actually tries (and fails) to install...
<details>
<code><pre>
D:\code\4node\node-gyp-1144$ npm i fsevents@0.3.4
npm http fetch GET 304 https://registry.npmjs.org/fsevents 239ms (from cache)
npm http fetch GET 304 https://registry.npmjs.org/nan 70ms (from cache)
npm http fetch GET 200 https://registry.npmjs.org/nan/-/nan-1.5.3.tgz 127ms
npm http fetch GET 200 https://registry.npmjs.org/fsevents/-/fsevents-0.3.4.tgz 130ms

> fsevents@0.3.4 install D:\code\4node\node-gyp-1144\node_modules\fsevents
> node-gyp rebuild

@CLINK_PROMPTD:\code\4node\node-gyp-1144\node_modules\fsevents$ if not defined npm_config_node_gyp (node "D:\bin\dev\node\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
Traceback (most recent call last):
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\gyp_main.py", line 16, in <module>
    sys.exit(gyp.script_main())
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 545, in script_main
    return main(sys.argv[1:])
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 538, in main
    return gyp_main(args)
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\__init__.py", line 523, in gyp_main
    generator.GenerateOutput(flat_list, targets, data, params)
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py", line 2019, in GenerateOutput
    sln_projects, project_objects, flat=msvs_version.FlatSolution())
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py", line 1778, in _GatherSolutionFolders
    return _DictsToFolders('', root, flat)
  File "D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\gyp\pylib\gyp\generator\msvs.py", line 1731, in _DictsToFolders
    for folder, contents in bucket.iteritems():
AttributeError: 'MSVSProject' object has no attribute 'iteritems'
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (D:\bin\dev\node\node_modules\npm\node_modules\node-gyp\lib\configure.js:336:16)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:194:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
gyp ERR! System Windows_NT 10.0.15063
gyp ERR! command "C:\\bin\\dev\\node\\node.exe" "D:\\bin\\dev\\node\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd D:\code\4node\node-gyp-1144\node_modules\fsevents
gyp ERR! node -v v7.10.0
gyp ERR! node-gyp -v v3.6.1
gyp ERR! not ok
npm WARN enoent ENOENT: no such file or directory, open 'D:\code\4node\node-gyp-1144\package.json'
npm WARN node-gyp-1144 No description
npm WARN node-gyp-1144 No repository field.
npm WARN node-gyp-1144 No README data
npm WARN node-gyp-1144 No license field.
npm WARN node-gyp-1144 Invalid dependency: windows-autoconf undefined

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! fsevents@0.3.4 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the fsevents@0.3.4 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\bin\var\npm-cache\_logs\2017-05-31T02_24_21_771Z-debug.log

D:\code\4node\node-gyp-1144$
</pre></code></details>

rolling back to `npm@4.6.1` I get a quick `EBADPLATFORM`

<details>
<code><pre>
D:\code\4node\node-gyp-1144$ npm i -g npm@4.6.1
npm http fetch GET 200 https://registry.npmjs.org/npm 463ms
npm http fetch GET 200 https://registry.npmjs.org/npm/-/npm-4.6.1.tgz 4496ms
c:\bin\dev\node\npm -> c:\bin\dev\node\node_modules\npm\bin\npm-cli.js
added 12 packages, removed 60 packages and updated 44 packages in 22.724s
<br>

D:\code\4node\node-gyp-1144$ npm i fsevents@0.3.4
npm http request GET https://registry.npmjs.org/fsevents
npm http 304 https://registry.npmjs.org/fsevents
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for fsevents@0.3.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"win32","arch":"x64"})
npm ERR! notsup Valid OS:    darwin
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   win32
npm ERR! notsup Actual Arch: x64

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\bin\var\npm-cache\_logs\2017-05-31T02_26_11_902Z-debug.log
</pre></code></details>

#### How can the CLI team reproduce the problem?
On windows in an appropriate directory run `npm i fsevents@0.3.4`

### supporting information:

 - `npm -v` prints: `5.0.0`
 - `node -v` prints: `7.10.0`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: `Windows`

**Comment by refack** [comment]
Cross-ref: https://github.com/nodejs/node-gyp/issues/1213

**Comment by zkat** [comment]
Awesome, thanks for finding this 👍

**Comment by oswaldofreitas** [comment]
```
> node install

node-pre-gyp info it worked if it ends with ok
node-pre-gyp verb cli [ '/usr/local/Cellar/node/8.0.0/bin/node',
node-pre-gyp verb cli   '/Users/me/projects/xxx/node_modules/fsevents/node_m
odules/.bin/node-pre-gyp',
node-pre-gyp verb cli   'install',
node-pre-gyp verb cli   '--fallback-to-build' ]
node-pre-gyp info using node-pre-gyp@https://registry.npmjs.org/node-pre-gyp/-/node-pre-
gyp-0.6.33.tgz
node-pre-gyp info using node@8.0.0 | darwin | x64
node-pre-gyp verb command install []
node-pre-gyp ERR! install error
node-pre-gyp ERR! stack TypeError: Cannot read property 'version' of null
...
```

macOS 10.12.5
node 8.0.0
npm 5.0.0

**Comment by jthegedus** [comment]
Thought I would also post as I use a different environment as others here. I use Bash on Ubuntu on Windows.

**Fails with**
Windows 10 1703
Bash on Ubuntu on Windows - `lsb_releae -a` => Ubuntu 16.04.2 LTS
Node 8.0.0
npm 5.0.1
```
Projects/Playground/gyp-node-8-test via ⬢  v8.0.0
➔ npm i -g firebase-tools
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
/home/jthegedus/.nvm/versions/node/v8.0.0/bin/firebase -> /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/bin/firebase

> grpc@1.2.4 install /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc
> node-pre-gyp install --fallback-to-build

node-pre-gyp info it worked if it ends with ok
node-pre-gyp verb cli [ '/home/jthegedus/.nvm/versions/node/v8.0.0/bin/node',
node-pre-gyp verb cli   '/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/node_modules/.bin/node-pre-gyp',
node-pre-gyp verb cli   'install',
node-pre-gyp verb cli   '--fallback-to-build' ]
node-pre-gyp info using node-pre-gyp@0.6.34
node-pre-gyp info using node@8.0.0 | linux | x64
node-pre-gyp verb command install []
node-pre-gyp info check checked for "/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary/grpc_node.node" (not found)
node-pre-gyp http GET https://storage.googleapis.com/grpc-precompiled-binaries/node/grpc/v1.2.4/node-v57-linux-x64.tar.gz
node-pre-gyp http 403 https://storage.googleapis.com/grpc-precompiled-binaries/node/grpc/v1.2.4/node-v57-linux-x64.tar.gz
node-pre-gyp ERR! Tried to download(403): https://storage.googleapis.com/grpc-precompiled-binaries/node/grpc/v1.2.4/node-v57-linux-x64.tar.gz
node-pre-gyp ERR! Pre-built binaries not found for grpc@1.2.4 and node@8.0.0 (node-v57 ABI) (falling back to source compile with node-gyp)
node-pre-gyp http 403 status code downloading tarball https://storage.googleapis.com/grpc-precompiled-binaries/node/grpc/v1.2.4/node-v57-linux-x64.tar.gz
node-pre-gyp verb command build [ 'rebuild' ]
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:483:19)
gyp ERR! stack     at PythonFinder.<anonymous> (/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:397:16)
gyp ERR! stack     at F (/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/which/which.js:68:16)
gyp ERR! stack     at E (/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/which/which.js:80:29)
gyp ERR! stack     at /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/which/which.js:89:16
gyp ERR! stack     at /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/which/node_modules/isexe/index.js:42:5
gyp ERR! stack     at /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/which/node_modules/isexe/mode.js:8:5
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:152:21)
gyp ERR! System Linux 4.4.0-43-Microsoft
gyp ERR! command "/home/jthegedus/.nvm/versions/node/v8.0.0/bin/node" "/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "--fallback-to-build" "--module=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary/grpc_node.node" "--module_name=grpc_node" "--module_path=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary"
gyp ERR! cwd /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc
gyp ERR! node -v v8.0.0
gyp ERR! node-gyp -v v3.6.1
gyp ERR! not ok
node-pre-gyp ERR! build error
node-pre-gyp ERR! stack Error: Failed to execute '/home/jthegedus/.nvm/versions/node/v8.0.0/bin/node /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary/grpc_node.node --module_name=grpc_node --module_path=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at emitTwo (events.js:125:13)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:213:7)
node-pre-gyp ERR! stack     at maybeClose (internal/child_process.js:887:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:208:5)
node-pre-gyp ERR! System Linux 4.4.0-43-Microsoft
node-pre-gyp ERR! command "/home/jthegedus/.nvm/versions/node/v8.0.0/bin/node" "/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc
node-pre-gyp ERR! node -v v8.0.0
node-pre-gyp ERR! node-pre-gyp -v v0.6.34
node-pre-gyp ERR! not ok
Failed to execute '/home/jthegedus/.nvm/versions/node/v8.0.0/bin/node /home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js configure --fallback-to-build --module=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary/grpc_node.node --module_name=grpc_node --module_path=/home/jthegedus/.nvm/versions/node/v8.0.0/lib/node_modules/firebase-tools/node_modules/grpc/src/node/extension_binary' (1)
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: grpc@1.2.4 (node_modules/firebase-tools/node_modules/grpc):
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: grpc@1.2.4 install: `node-pre-gyp install --fallback-to-build`
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: Exit status 1

added 148 packages and updated 1 package in 74.205s

Projects/Playground/gyp-node-8-test via ⬢  v8.0.0
➔ 
```


**Works with**
Windows 10 1703
Bash on Ubuntu on Windows - `lsb_releae -a` => Ubuntu 16.04.2 LTS
Node 6.10.3
npm 3.10.10
```
Projects/Playground/gyp-node-8-test via ⬢  v8.0.0
➔ nvm use --lts
Now using node v6.10.3 (npm v3.10.10)

Projects/Playground/gyp-node-8-test via ⬢  v6.10.3
➔ npm i -g firebase-tools
npm WARN deprecated node-uuid@1.4.8: Use uuid module instead
/home/jthegedus/.nvm/versions/node/v6.10.3/bin/firebase -> /home/jthegedus/.nvm/versions/node/v6.10.3/lib/node_modules/firebase-tools/bin/firebase
/home/jthegedus/.nvm/versions/node/v6.10.3/lib
└─┬ firebase-tools@3.9.0
  └─┬ es6-set@0.1.5
    └── es5-ext@0.10.22


Projects/Playground/gyp-node-8-test via ⬢  v6.10.3
➔
```

**Comment by zkat** [comment]
So in order to fix this, I gotta comb through and figure out where it was that `npm@4` used this before, and then see what a good home for the new code is. There's various places where this check _could_ happen but I gotta talk it through with @iarna 'cause it can cause different behavior depending on where it lands.

**Comment by jaredbeck** [comment]
Downgrading npm with `npm i npm@4.6.1 -g` worked for me. Obviously not a permanent solution, but hope it helps someone.

**Comment by gaearon** [comment]
Seemingly, another instance of this issue: https://github.com/facebookincubator/create-react-app/issues/2558.
Affects everyone using Create React App on Windows.

**Comment by reeddunkle** [comment]
A teammate found a solution that works with npm v5.0.x:

1. delete the `node_modules` directory
2. delete the `package-lock.json` file
3. `npm i`

**Comment by zkat** [comment]
@gaearon @refack https://github.com/npm/npm/pull/17552 should fix this. It'll land in the canary as soon as travis is happy with it, and it will be included in the 5.1.0 release next week (probably ~wednesday)

**Comment by zkat** [comment]
Closing this, since #17552 has landed in `release-next`. It is available in the canary as of `npmc@5.0.4-canary.13`, and will be available in mainline npm next Wednesday. Thanks all!
